### DISCUSSION TITLE
Towards reusable thread-group code

### DISCUSSION THREAD

**Discussion by AntonioSun** [smicyk/groovy-jmeter#121] [Ideas]
I know the DSL [groovy-jmeter](https://github.com/smicyk/groovy-jmeter) provides currently is focus on wrapping over JMeter ones, but am hoping it can go one level up to achieve the following --

In our testing, we have standalone tests that focus on one specific area, and mixed load test that put them _all_ together. The current situation is, both in [groovy-jmeter](https://github.com/smicyk/groovy-jmeter) and JMeter, it's rather difficult to reuse the thread-group code for both cases, as the load design/requirement is very different. As the result, we have to break the following natually grouped code into two parts, the `group` line and its content:

```groovy
        // define group of 1000 users with ramp up period 300 seconds
        group(users: 1000, rampUp: 300) {
            // add cookie manager for HTTP requests
. . .
```

Such break-up doesn't feel natual. the `group` line and its content are seprated.

If somehow we can design a DSL that allows the insert code file can have the `group` line within it, and pass every prameter that `group` can accept to it, it'll looks ideal, ..., but, having written it all out, I'm now wondering if it is worth the trouble 😊😊...

**Comment by smicyk** [comment]
Hi, it seems valid request. So as understand you what you would like to have is something like:

```groovy
// test.groovy

start {
   // simple case
   insert 'fragments/group.groovy', properties: [ users: 1, rampUp: 10]

   // load test
   insert 'fragments/group.groovy', properties: [users: 1000, rampUp: 300]
}

// fragments/group.groovy
fragment {
    group users: users, rampUp: rampUp, {
         cookies()
         simple {
               http 'GET sample.com'
         }
    }
}

```

This should be quite doable but you can try to do it in some other way e.g.

```groovy
// test.groovy

@GrabConfig(systemClassLoader=true)                                             
@Grab('net.simonix.scripts:groovy-jmeter')                                     
                                                                               
@groovy.transform.BaseScript net.simonix.dsl.jmeter.TestScript script          
                                                                               

start {
    plan 'Test plan', {
        group users: 1, rampUp: 10, {
            insert 'fragments/group.groovy'
        }
        
        group users: 1000, rampUp: 300, {
            insert 'fragments/group.groovy'
        }    
    }
}

// fragments/group.groovy
fragment {
    cookies()
    simple {
         http 'GET sample.com'
    }
}
```

Hope it helps

**Reply by AntonioSun** [reply]
> This should be quite doable but you can try to do it in some other way

Yes, this is exactly what I'm doing now, which I can continue doing. 

Your proposed solution looks fabulous, even better than I had imagined. Looking forward to it!

**Reply by smicyk** [reply]
Hi, with this #126 the new variables property added to insert keyword so you can pass variables to fragment.

**Reply by AntonioSun** [reply]
👏👏👏🎉🎉🎉
Time to cut a new release maybe?

**Reply by smicyk** [reply]
Done. [v0.22.0](https://github.com/smicyk/groovy-jmeter/releases/tag/v0.22.0)

Should be available in maven repos in 1-2 days.
